Make C++ RpcAgent::currentRPCAgent_ the source of truth of current RPC Agent

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32635 Use the C++ current RpcAgent pointer to eliminate the unnecessary argument passing from Python world
* **#32633 Make C++ RpcAgent::currentRPCAgent_ the source of truth of current RPC Agent**

There were 2 sources of current RPC agent.

- One is in Python world, `torch.distributedrpc.api._agent`.
- The other is in C++ world, `RpcAgent::defaultRpcAgent_`

Setting Python `_agent` to `None`, does not necessarily reset the C++ `defaultRpcAgent_` to `nullptr`.

i.e.
```
 torch.distributedrpc.api._agent = None
```
does not translate to
```
RpcAgent::defaultRpcAgent_ = nullptr
```

This PR is to remove this ambiguity, and use the C++ pointer as source of truth.

The solution is to leverage a pybind11 behavior that it implicitly casts C++ `shared_ptr<RpcAgent>(nullptr)` to Python `None`.

Differential Revision: [D5733066](https://our.internmc.facebook.com/intern/diff/D5733066/)

xush6528 committed:
Make C++ RpcAgent::defaultRPCAgent() the source of truth of current RPC Agent

There were sources of cuirrent RPC agent.

- One is in Python world, `torch.distributedrpc.api._agent`.
- The other is in C++ world, `RpcAgent::defaultRpcAgent_`

Setting Python `_agent` to `None`, does not necessarily reset the C++ `defaultRpcAgent_` to `nullptr`.

i.e.
```
 torch.distributedrpc.api._agent = None
```
does not translate to
```
RpcAgent::defaultRpcAgent_ = nullptr
```

This PR is to remove this ambiguity, and use the C++ pointer as source of truth.

The solution is to leverage a pybind11 behavior that it implicitly casts C++ `shared_ptr<RpcAgent>(nullptr)` to Python `None`.

Differential Revision: [D5733066](https://our.internmc.facebook.com/intern/diff/D5733066/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit fe88e413:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=6ad9e5c70d1f496b9eaca191ec18e26c106c6436&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 6ad9e5c7 since Jan 25
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  Since your merge base is older than `viable/strict`, run these commands:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=fe88e41316d73793c6ea9dcfb8a5c072129aa0da).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_macos_10_13_py3_test](https://dr.pytorch.org/build-details.html?build_id=91073714) from Jan 25 until Jan 25



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -253,6 +253,16 @@ If the future completes with an error, an exception is thrown.
           py::call_guard<py::gil_scoped_release>());
 
+  module.def(
+      "get_current_rpc_agent",
```
any reason this does not start with an underscore?

xush6528 replied:
I thought it's a public API. Looking thorough call sites again. I think it's only used in rpc_test.py and some rpc starter library code. makes sense to keep it conservative and add an underscore for now.

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -24,6 +25,7 @@
     _invoke_rpc_script,
     _set_rpc_timeout,
     _start_rpc_agent,
+    _reset_current_rpc_agent,
```
this should be above _set_rpc_timeout?

xush6528 replied:
Right. I though formatter could order then, but no.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -253,6 +253,16 @@ If the future completes with an error, an exception is thrown.
+  module.def(
+      "get_current_rpc_agent",
+      &RpcAgent::getDefaultRpcAgent,
+      py::call_guard<py::gil_scoped_release>());
```
Should we not release GIL on these two functions to avoid ctx switch as they are very light?

xush6528 replied:
Make sense to me, let me remove.

mrshenli reviewed (approved):
LGTM! Thanks!

xush6528 committed:
Update on "Make C++ RpcAgent::currentRPCAgent_ the source of truth of current RPC Agent"


There were 2 sources of current RPC agent.

- One is in Python world, `torch.distributedrpc.api._agent`.
- The other is in C++ world, `RpcAgent::defaultRpcAgent_`

Setting Python `_agent` to `None`, does not necessarily reset the C++ `defaultRpcAgent_` to `nullptr`.

i.e.
```
 torch.distributedrpc.api._agent = None
```
does not translate to
```
RpcAgent::defaultRpcAgent_ = nullptr
```

This PR is to remove this ambiguity, and use the C++ pointer as source of truth.

The solution is to leverage a pybind11 behavior that it implicitly casts C++ `shared_ptr<RpcAgent>(nullptr)` to Python `None`.

Differential Revision: [D5733066](https://our.internmc.facebook.com/intern/diff/D5733066/)

[ghstack-poisoned]

xush6528 commented:
- Rename `defaultRpcAgent_` to `currentRpcAgent_`.
- Add a `isCurrentRpcAgentSet` API, so that we can assert that `getCurrentRpcAgent` never returns a nullptr.

xush6528 committed:
Update on "Make C++ RpcAgent::currentRPCAgent_ the source of truth of current RPC Agent"


There were 2 sources of current RPC agent.

- One is in Python world, `torch.distributedrpc.api._agent`.
- The other is in C++ world, `RpcAgent::defaultRpcAgent_`

Setting Python `_agent` to `None`, does not necessarily reset the C++ `defaultRpcAgent_` to `nullptr`.

i.e.
```
 torch.distributedrpc.api._agent = None
```
does not translate to
```
RpcAgent::defaultRpcAgent_ = nullptr
```

This PR is to remove this ambiguity, and use the C++ pointer as source of truth.

The solution is to leverage a pybind11 behavior that it implicitly casts C++ `shared_ptr<RpcAgent>(nullptr)` to Python `None`.

Differential Revision: [D5733066](https://our.internmc.facebook.com/intern/diff/D5733066/)

[ghstack-poisoned]

xush6528 committed:
Update on "Make C++ RpcAgent::currentRPCAgent_ the source of truth of current RPC Agent"


There were 2 sources of current RPC agent.

- One is in Python world, `torch.distributedrpc.api._agent`.
- The other is in C++ world, `RpcAgent::defaultRpcAgent_`

Setting Python `_agent` to `None`, does not necessarily reset the C++ `defaultRpcAgent_` to `nullptr`.

i.e.
```
 torch.distributedrpc.api._agent = None
```
does not translate to
```
RpcAgent::defaultRpcAgent_ = nullptr
```

This PR is to remove this ambiguity, and use the C++ pointer as source of truth.

The solution is to leverage a pybind11 behavior that it implicitly casts C++ `shared_ptr<RpcAgent>(nullptr)` to Python `None`.

Differential Revision: [D5733066](https://our.internmc.facebook.com/intern/diff/D5733066/)

[ghstack-poisoned]

xush6528 committed:
Update on "Make C++ RpcAgent::currentRPCAgent_ the source of truth of current RPC Agent"


There were 2 sources of current RPC agent.

- One is in Python world, `torch.distributedrpc.api._agent`.
- The other is in C++ world, `RpcAgent::defaultRpcAgent_`

Setting Python `_agent` to `None`, does not necessarily reset the C++ `defaultRpcAgent_` to `nullptr`.

i.e.
```
 torch.distributedrpc.api._agent = None
```
does not translate to
```
RpcAgent::defaultRpcAgent_ = nullptr
```

This PR is to remove this ambiguity, and use the C++ pointer as source of truth.

The solution is to leverage a pybind11 behavior that it implicitly casts C++ `shared_ptr<RpcAgent>(nullptr)` to Python `None`.

Differential Revision: [D5733066](https://our.internmc.facebook.com/intern/diff/D5733066/)

[ghstack-poisoned]

xush6528 committed:
Update on "Make C++ RpcAgent::currentRPCAgent_ the source of truth of current RPC Agent"


There were 2 sources of current RPC agent.

- One is in Python world, `torch.distributedrpc.api._agent`.
- The other is in C++ world, `RpcAgent::defaultRpcAgent_`

Setting Python `_agent` to `None`, does not necessarily reset the C++ `defaultRpcAgent_` to `nullptr`.

i.e.
```
 torch.distributedrpc.api._agent = None
```
does not translate to
```
RpcAgent::defaultRpcAgent_ = nullptr
```

This PR is to remove this ambiguity, and use the C++ pointer as source of truth.

The solution is to leverage a pybind11 behavior that it implicitly casts C++ `shared_ptr<RpcAgent>(nullptr)` to Python `None`.

Differential Revision: [D5733066](https://our.internmc.facebook.com/intern/diff/D5733066/)

[ghstack-poisoned]
